### ISSUE TITLE
Issue with NaNs

### ISSUE THREAD

**Issue by ericmjl** [ElArkk/jax-unirep#37]
I tried debugging and found that the issue is most likely in the prediction of y_hat in the evotune_loss() function. The y_hat tensor aka "predictions" can contain NaNs --> loss will be NaN, as well as values > 1 --> loss will be negative. I also found cases where y_hat tensor has large negative values. This to me seems like an issue with softmax. Unfortunately I don't have a solid enough understanding of jax's vmap and stax.serial predict in order to debug this, and was wondering if you could help.

_Originally posted by @ivanjayapurna in https://github.com/ElArkk/jax-unirep/pull/36#issuecomment-615486320_

**Comment by ericmjl** [comment]
> How thoroughly have you tested the evotuning function (before any modifications I made)? I was experimenting with it and can't seem to get reasonable losses. I've seen cases where the cross entropy losses are NaN as well as where they are negative values.

We made it work quite a number of times, but also had to struggle through NaNs before, and they showed up at different times depending on the task.

Some things to maybe check for, let me know if you've already checked them before:

1. Drop step size to an unreasonably small size, like 2 orders of magnitude smaller than the step size that gives you NaNs.
1. If you are also evotuning end-to-end with a target task, make sure that your target task is a column vector.

As for the cross entropy loss, we did change it from a [numerically unstable version](https://github.com/ElArkk/jax-unirep/blob/e29b35f83c81fc972654735567b9feac4a26ebf1/jax_unirep/losses.py#L4) to the [numerically stable version](https://github.com/ElArkk/jax-unirep/blob/e29b35f83c81fc972654735567b9feac4a26ebf1/jax_unirep/losses.py#L15) (also making it negative along the way). I think we looked at the TensorFlow version as a reference for how to make it work, but I now regret not putting in the docs the URL to the TF docs that we referenced.

If you'd like, I'm happy to jump on a call on Monday to see what might be going on. Given you're in California, and I have meetings in the morning till 2 pm, if you've got time I'm happy to call after 2 pm ET/10 am PT. (Miss the west coast!) This is the kind of thing where debugging in-person with the goal of making it robust long-term might be easier than discussing via text. @ElArkk is in Basel, so 2 PM ET is 8 PM CET. Anyways, @ivanjayapurna lemme know what you'd like to do.

**Comment by ericmjl** [comment]
from @ivanjayapurna 

> A couple immediate comments:
> 1. Could you clarify what exactly you're calling a "target task" that should be a column vector?
> 1. Given the numerically stable loss function with the negative at the front, assuming y and y_hat are correctly generated as values between 0 - 1 (which is definitely true for y as it is just one-hot, and should be true for y_hat if the softmax layer is working properly - but currently I don't believe it is), then loss values should always be positive right? I'm not sure how to interpret negative loss values in this situation other than y_hat gave probabilities > 1.


Here's my replies to your questions:

> Could you clarify what exactly you're calling a "target task" that should be a column vector?

"target task" refers to the thing you're trying to predict, e.g. protein melting point. If you're doing evotuning without a target task, then the target task is the prediction of the next amino acid letter. If you're doing end-to-end fitting, e.g. training the UniRep weights alongside feed-forward neural network weights to predict protein melting point, then the protein melting point `y` should be a column vector rather than a row vector.

This is why: if `y` is of shape `(n_samples,)`, while `y_hat` is of shape `(n_samples, 1)`, then I think NumPy's broadcasting rules will coerce `y * np.log(y_hat)` to be of shape `(n_samples, n_samples)`, and thus the calculation of the loss function will be messed up.

That said, if you're evotuning only, with the learning task being "predict next amino acid letter", then the code we have should set up the evotuning pairs such that the shapes are all correct. The output should be of shape `(n_positions - 1, 26)`. `n_positions - 1` because we are predicting next letter. `26` because there are 26 one-hot options. (Or it could be 25, I'm going off memory to not interrupt writing flow.)

> Given the numerically stable loss function with the negative at the front, assuming y and y_hat are correctly generated as values between 0 - 1 (which is definitely true for y as it is just one-hot, and should be true for y_hat if the softmax layer is working properly - but currently I don't believe it is), then loss values should always be positive right? I'm not sure how to interpret negative loss values in this situation other than y_hat gave probabilities > 1.

This is a good thing to check. Do you see invalid y_hat values with simulated data as well? Or do you see it only with the data that you're trying to use? If so, is it possible to narrow down the situations where you're seeing NaNs? One of the important things around debugging computer programs is being able to produce a minimal reproducible example (a "[reprex](https://pypi.org/project/reprexpy/)") that can be copy/pasted. It's frustrating, yes, but by investing the time to isolate the problem from the larger compute environment, we can see whether it's due to the code only or if there's interactions with the compute environment that are lurking around (e.g. system-wide floating point settings have bitten me before). The reprex link above will take you to a pip-installable tool that you could consider installing into every project environment to help you with debugging :smile:. 

One thing to try is to write a function that checks for NaNs or invalid values on every loop, and then break the loop. Always check the forward pass of the model for NaNs. You should also check the input data, i.e. the result from the `evotuning_pairs` function, for invalid data. I think we have a software test for it, but if we missed something, that's a bug to fix and test against in the test suite.

**Comment by ericmjl** [comment]
@ivanjayapurna I just sent you a calendar invite (had to sleuth a bit for an email - and I would be hesitant to ask you for it here as well). Feel free to counter-propose if it clashes.

**Comment by ivanjayapurna** [comment]
Accepted the calendar invite and am still down to call because I'm sure there are still other problems that will come up between now and then, but I think I fixed the problem! I tested my hypothesis that the softmax layer wasn't working by manually implementing a softmax layer - and it works (gives predicted probabilities between 0-1 and positive losses). I will try to see if I can find cases of NaN values with my new softmax layer.

**Comment by ElArkk** [comment]
@ivanjayapurna that would be great news!

Before we switched to the `jax` implementation of softmax, we actually did use a custom one as well. It didn't work properly in the beginning (producing NaN's), and so we tried quite a number of variations on it, never achieving that it _always_ worked. So you should really test it rigorously under different training regimes and for prolonged periods of time. If no errors show up, then that's perfect :)

**Comment by ElArkk** [comment]
@ivanjayapurna @ericmjl update:

I just pulled @ivanjayapurna 's branch down and let his `evotune` as well as the current implementation run on some test sequences.
The current implementation gave increasingly negative losses, but no NaN's. I sadly don't remember whether we encountered negative losses before we made the switch to using `stax` (that might have changed stuff too)
Ivan's implementation gave nice positive losses, but then after 3 epochs, NaN's popped up. If I look at his `softmax`, I think the most likely problem is that it doesn't have protection against overflow caused by big `y_hat`'s. `jax`'s implementation actually guards against this by subtracting `y_hat.max` from all `y_hat`'s (see https://github.com/google/jax/blob/94178d98a055eca94f836ecc2ca7a690c5186b66/jax/nn/functions.py#L213)

**Comment by ivanjayapurna** [comment]
@ElArkk  Good catch, in my trials with new softmax I never tested beyond 3 epochs. I just added the guard to protect against overflow to my softmax and it seems to still give good positive, decreasing losses! Will push it now - hopefully that should solve the NaN issues.

**Comment by ivanjayapurna** [comment]
@ElArkk  Good catch, in my trials with new softmax I never tested beyond 3 epochs. I just added the guard to protect against overflow to my softmax and it seems to still give good positive, decreasing losses! Will push it now - hopefully that should solve the NaN issues.

**Comment by jmahenriques** [comment]
Apologies for crashing the issue like this, but this is what you're referring to, right? 

`Setting status of trial#0 as TrialState.FAIL because the objective function returned nan.`

I am currently trying out your code, so if there's anything I can do to help please let me know. I have absolutely no experience with JAX, but maybe I can be of use in some capacity since I am very interested in using your code due to its simplicity of use when compared to other alternatives.

I would also like to discuss evotuning performance and GPU usage, but that's not appropriate for this issue thread. Maybe we can communicate through email or some other medium?

**Comment by ericmjl** [comment]
@jmahenriques welcome to the issue tracker! 

The issue tracker is not a problem for this, though it's best if the issues are kept isolated and on-topic. For performance, let's go to a new issue.

**Comment by ElArkk** [comment]
With Ivan's PR merged and no issues with NaN's being reported up to now, I'm going to close this issue for now :)
